book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,211 @@
+
+We need the raw representation to Serialize/Deserialize the transaction however Zebra uses its own representation internally.
+
+V5 transactions are the only ones that will support orchard transactions with `Orchard` data types.
+
+V4 and V5 transactions both support sapling data but they are implemented differently. By this reason we need to split sapling data types into V4 and V5.
+
+The order of some of the fields changed from V4 to V5. For example the `lock_time` and `expiry_height` were moved above the transparent inputs and outputs.
+
+All of the changes proposed in this document are only to the `zebra-chain` crate.
+
+To highlight changes all document comments from the code snippets in the [reference section](#reference-level-explanation) were removed.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We start by looking how a V4(already implemented) transaction is represented in Zebra.
+
+This transaction version is specified by the protocol in the first table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<ShieldedData>,
+}
+```
+
+Currently the `ShieldedData` type is defined in `zebra-chain/src/transaction/shielded_data.rs` as follows:
+
+```
+pub struct ShieldedData {
+    pub first: Either<Spend, Output>,
+    pub rest_spends: Vec<Spend>,
+    pub rest_outputs: Vec<Output>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+We know by protocol(2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    sapling_value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<SaplingV4ShieldedData>,
+}
+```
+
+The only change above from the current implementation is the `shielded_data` field type, changed from  `Option<ShieldedData>` to `Option<SaplingV4ShieldedData>`.
+
+`ShieldedData` is currently defined and implemented in `src/transaction/shielded_data.rs`. We propose to change `ShieldedData` into `SaplingV4ShieldedData` and to add `SaplingV5ShieldedData` into the same file.
```

Comment:
Here and in other places below in the document we have 2 options. We include the new V4 and V4 types in the same file or we create new files for each of them.

In this case `SaplingV4ShieldedData` and `SaplingV5ShieldedData` are living in the same file `src/transaction/shielded_data.rs` and so the methods, implementations, serializations, etc.

The other option is to have `src/transaction/shielded_data_v4.rs` with all related to `SaplingV4ShieldedData` on it and `src/transaction/shielded_data_v5.rs` with  `SaplingV5ShieldedData`.

I dont have a strong opinion for any of the 2, both options require the same amount of duplicated code but the current version of the RFC is going with the first option(have V4 and V5 types in the same file).

A similar situation is happening in sapling spend and outputs where the current version of the RFC is placing the 2 types into the same already existing file.